### ISSUE TITLE
Centralized Logging

### ISSUE THREAD

**Issue by Cahllagerfeld** [EddieHubCommunity/api#60]
When growing the Application we might need some Monitoring about Performance, Failures or Errors. We could potentially implement centralized logging to Azure App Insights or Sentry. 
If we do so, we could also implement this to the different Frontends like EddieBot or a Website

**Comment by eddiejaoude** [comment]
Great idea. We will be deploying to a Digital Ocean Kubenertes cluster, I wonder if they offer something or if we can add a plugin the the cluster

**Comment by Cahllagerfeld** [comment]
Sounds good!

**Comment by Cahllagerfeld** [comment]
Lets do this next.
I think we should collect some ideas of how we want to Log.
- Do we want to use an external monitoring tool?
- What do we want to log?
- Do we want Log-Files?
- Where do we store Log-Files in case we want them?

**Comment by Cahllagerfeld** [comment]
> 
> 
> Lets do this next.
> I think we should collect some ideas of how we want to Log.
> 
>     * Do we want to use an external monitoring tool?
> 
>     * What do we want to log?
> 
>     * Do we want Log-Files?
> 
>     * Where do we store Log-Files in case we want them?

I think we should definitly log every Http-Execption `>400`

**Comment by eddiejaoude** [comment]
It would be great if we could have a `DEBUG=true` mode where we log every request

I think we keep it simple for now, can we log to the console, but with a config change then able to log somewhere else?

**Comment by Cahllagerfeld** [comment]
Yes, I think thats possible.
I've done sth similar to this this before. We did it with `winston` and checked for environmet variables when creating the different transports. 
I think writing a LogginInterceptor is best if we want to log every Request.

**Comment by Cahllagerfeld** [comment]
> 
> 
> Lets do this next.
> I think we should collect some ideas of how we want to Log.
> 
>     1. Do we want to use an external monitoring tool?
> 
>     2. What do we want to log?
> 
>     3. Do we want Log-Files?
> 
>     4. Where do we store Log-Files in case we want them?

Sorry that I was a little exuberant just now, but these were the thoughts we had when implementing logging 😅

**Comment by eddiejaoude** [comment]
lol no problem 🤓 

Should we start with using the vanilla NestJS logger? https://docs.nestjs.com/techniques/logger#using-the-logger-for-application-logging

**Comment by Cahllagerfeld** [comment]
This is the question I also have :D 
I think for now the default Logger is fine.
If we want to expand logging we probably have to use a logger-package then.

**Comment by eddiejaoude** [comment]
Hopefully if / when we want to expand it will be in a single place to refactor and everything else will continue to work

**Comment by Cahllagerfeld** [comment]
Which Logformat do we want?
```ts
this.logger.log(`${req.method}] ${req.originalUrl}`);
```
and do we also want to log the response-status code, that is sent back to the client?

**Comment by eddiejaoude** [comment]
Good question, I think the response status code would be useful 👍 

Format looks good, is there a missing closing bracket?

```diff
- this.logger.log(`${req.method}] ${req.originalUrl}`);
+ this.logger.log(`[${req.method}] ${req.originalUrl}`);
```

What about the request data? Can we log that too or is it too much data?

**Comment by Cahllagerfeld** [comment]
Oh I probably made a mistake when typing it in here 😅 
I had to lay the logic into a middleware otherwise 401 wouldnt be logged, as Guards are called before interceptors.
The logic can be turned off by settings environment variable `DEBUG=false` or leaving `DEBUG` undefined.
![grafik](https://user-images.githubusercontent.com/43843195/115776499-78bc9e00-a3b4-11eb-9340-f8e887dc607a.png)
I just found a little mistake which I have to fix tomorrow.
